### ISSUE TITLE
Step API - various utility functions for modules and UIs to reference step state

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer#242]
While working on #214 I realize there are a lot of common needs that one has when writing a module or a UI. Writing a complex interface/module is a great way to discover what extra utility methods we really need to make life easier for module developers.

Let's make a list, and we can work to add them as we go -- this is a list of standard values or functions we should make available both to a given module's instantiation as a "step" -- and to the UI that it generates and updates:

## Step

* [ ] the input image `src` - as soon as it's first passed in - see implementation in #214's custom UI
* [ ] the output image `src` - as soon as it's first calculated
* [ ] listeners for the above events? (`onDraw` for the first, `onComplete` for the second?)
* [ ] various properties of the above - `width`, `height`, `size`, `format`, or utilities to calculate them - `getWidth()`, `getHeight()` etc
* [ ] a reference to the previous step - `getPreviousStep()`
* [ ] a reference to the next step - `getNextStep()`
* [ ] a reference to the current step's `index` -- probably generated, like `getIndex()`
* [ ] a reference to any given step's generated metadata (like from QRCode module)

## Module

* the module name & description - `module.name`?

## Environment

* `inBrowser` (done, could be more consistent)
* `hasWebGl` (after #216 this will be important to distinguish simply from a browser)

### UI

These ought to be provided by the UI -- I'm focused on HTML right now, but let's ensure the default HTML UI sets a good example:

* [ ] unique selector to the "step" element -- in default UI - `step.element`
* [ ] the displayed image preview (in the default UI this is an instance of the actual image) - `step.imgElement` - believe this exists already but needs documentation
* [ ] link to the "details" or "settings" area - although this varies with UI and could be calculated from the `step.element` selector

**Comment by jywarren** [comment]
Made lots of small edits to this - worth looking over, @tech4GT !! Thanks!!

**Comment by tech4GT** [comment]
@jywarren going through this now😀

**Comment by jywarren** [comment]
yes, i agree!

Anything to add here that we've learned is useful?

**Comment by tech4GT** [comment]
@jywarren can you please explain the first and second point in step? what do we exactly mean by having it as soon as its calculated and passed??

**Comment by jywarren** [comment]
Just that it would be nice to, from inside the step, be able to refer back to the untouched original input image, and, after it's been run once, to the current (or most recent) output (for example to see if our calculated output is different from last time it was run?).

**Comment by tech4GT** [comment]
@jywarren Oh okay I got it so we can have functions like
```js
function getSource(){
}
```
but I still don't get the second point, do we want to have the outputs stored in an array, each time the step is run?

**Comment by jywarren** [comment]
I think the output is less needed. In theory you can ask the next step for
its input.

**Comment by tech4GT** [comment]
So @jywarren should I not do the second one?

**Comment by jywarren** [comment]
Yeah, i guess so, at least not for now. This is a kind of "wishlist" so we
can identify the most helpful items to implement in our first more
formalized Step API.

**Comment by tech4GT** [comment]
Oh i got it!! Actually i was seeing in terms of progress percentage of the core work milestone so we can implement the ones we need right now and then maybe remove this from the core work and keep breaking out separate issues out of this as we need them, what say?

**Comment by tech4GT** [comment]
@jywarren Can you please help me with that? I think we should handle UI later with the rest of the UI improvements so leaves us with step, can you please mark the ones we should do in step?

**Comment by jywarren** [comment]
I re-ordered them - I agree about the UI ones -- and otherwise i think the output src is really the only one we can leave for later. It won't even be too hard if you can get access to the next step, and it's input.

**Comment by jywarren** [comment]
UI ones are nice to break out into a separate issue -- all related

**Comment by jywarren** [comment]
Hi, for the listeners, I guess I am thinking of ways that in the step setup, you could create functions that would run upon specific triggers. We currently do this as part of the UI for some steps, but you could imagine this happening in the step itself for non-UI reasons too, perhaps? 

https://github.com/publiclab/image-sequencer/blob/master/src/ui/UserInterface.js

Maybe events for `onUpdate` for when settings change... can you think of others? And what do you think about these being pulled back from the UI into the main step API, for using outside the UI context? I'm not sure about this, so interested to hear your thoughts!

**Comment by tech4GT** [comment]
@jywarren yeah this could make sense for when options are updated and when the module is removed maybe, I think we should implement this separate from the UI api, since UI api handles events on sequencer and this new api will handle events on the module itself!!

**Comment by tech4GT** [comment]
All of core work on this has been done, taking this out of the milestone core work now!!
